src/main/java/com/birdie/kafka/connect/json/SchemaTransformer.java

```diff
@@ -191,56 +205,59 @@ public SchemaBuilder unionSchemas(Schema ...schemas) {
             throw new IllegalArgumentException("We can't union-ize an empty list of schemas.");
         }
 
-        List<String> types = List.of(schemas)
-                .stream()
-                .map(schema -> schema.type().toString())
-                .distinct()
-                .collect(Collectors.toList());
+        Schema.Type type = schemas[0].type();
 
-        if (types.size() != 1) {
-            throw new IllegalArgumentException("We can only union schemas of the same type together. Found: " + String.join(",", types));
-        }
+        // Array-specific values we care about
+        Schema.Type valueType = null;
+        Schema[] valueSchemas = new Schema[schemas.length];
 
-        SchemaBuilder schemaBuilder;
-        Schema.Type type = Schema.Type.valueOf(types.get(0));
+        // Struct-specific values we care about
+        Map<String, List<Schema>> fieldsByName = new HashMap<>();
 
-        if (type.equals(Schema.Type.ARRAY)) {
-            List<String> valueTypes = List.of(schemas)
-                    .stream()
-                    .map(schema -> schema.valueSchema().type().toString())
-                    .distinct()
-                    .collect(Collectors.toList());
-
-            if (valueTypes.size() != 1) {
-                throw new IllegalArgumentException("We can only union array schemas of the same value type together. Found: " + String.join(",", valueTypes));
+        for (int i = 0; i < schemas.length; i++) {
+            Schema schema = schemas[i];
+
+            if (!schema.type().equals(type)) {
+                throw new IllegalArgumentException("We can only union schemas of the same type together. Found: " + type + " and " + schema.type());
+            }
+
+            if (schema.type().equals(Schema.Type.ARRAY)) {
+                if (valueType == null) {
+                    valueType = schema.valueSchema().type();
+                } else if (!valueType.equals(schema.valueSchema().type())) {
+                    throw new IllegalArgumentException("We can only union array schemas of the same value type together. Found: " + valueType + " and "+ schema.valueSchema().type());
```

Comment:
Do we expect to hit these use cases much?

Reply:
Yup yup it happens 😅

Reply:
The assumption being that we _do_ create a new schema at this point? How do we keep track of that, or do we just let it fail?

Reply:
Yes exactly, there is another part of the code which creates an alternative schema. In practices, there is a list of schemas it tries to unify with, not just one.